lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx

```diff
@@ -0,0 +1,477 @@
+{
+  return permutationSize_;
+}
+
+/** Compute p-value with permutation */
+void HSICEstimatorImplementation::computePValuesPermutation() const
+{
+  SquareMatrix Wobs(computeWeightMatrix(outputSample_));
+  PValuesPermutation_ = Point(inputDimension_);
+
+  for(UnsignedInteger dim = 0; dim < inputDimension_; ++dim)
+  {
+    Sample xdim(inputSample_.getMarginal(dim));
+    Scalar HSIC_obs = computeHSICIndex(xdim, outputSample_, covarianceList_[dim], covarianceList_[inputDimension_], Wobs);
+
+    UnsignedInteger count = 0;
+    for( UnsignedInteger b = 0; b < permutationSize_; ++b)
+    {
+      Sample Yp = shuffledCopy(outputSample_);
+      SquareMatrix W(computeWeightMatrix(Yp));
+      Scalar HSIC_loc = computeHSICIndex(xdim, Yp, covarianceList_[dim], covarianceList_[inputDimension_], W);
+      if( HSIC_loc > HSIC_obs) count += 1;
+    }
+
+    /* p-value by permutation */
+    PValuesPermutation_[dim] = count * 1.0 / (permutationSize_ + 1) ;
+  }
+  isAlreadyComputedPValuesPermutation_ = true ;
+}
+
+/** Get the HSIC indices */
+Point HSICEstimatorImplementation::getHSICIndices() const
+{
+  if(!(isAlreadyComputedIndices_))
+  {
+    computeIndices();
+    isAlreadyComputedIndices_ = true ; 
+  }
+  return HSIC_XY_;
+}
+
+/** Get the R2-HSIC indices */
+Point HSICEstimatorImplementation::getR2HSICIndices() const
+{
+  if(!(isAlreadyComputedIndices_))
+  {
+    computeIndices();
+    isAlreadyComputedIndices_ = true ;
+  }
+  return R2HSICIndices_;
+}
+
+/** Get the p-values by permutation */
+Point HSICEstimatorImplementation::getPValuesPermutation() const
+{
+  if( !(isAlreadyComputedPValuesPermutation_))
+  {
+    computePValuesPermutation();
+    isAlreadyComputedPValuesPermutation_ = true ;
+  }
+  return PValuesPermutation_;
+}
+
+/** Draw the HSIC indices */
+Graph HSICEstimatorImplementation::drawValues(const Point &values, String &title) const
+{
+
+  if (values.getDimension() == 0) throw InvalidArgumentException(HERE) << "Error: cannot draw cloud based on empty data.";
+  Graph graph(title, "Input marginal number", "", true, "");
+
+  /* Define cloud */
+  Sample data(values.getDimension(), 2);
+  for (UnsignedInteger k = 0; k < values.getDimension(); ++k)
+  {
+    data(k, 0) = k + 1.0;
+    data(k, 1) = values[k];
+  }
+
+  Cloud cloud(data, "red", "circle", "");
+  graph.add(cloud);
+
+  /* Min & max indices values */
+  const Scalar minInd = data.getMin()[1];
+  const Scalar maxInd = data.getMax()[1];
+
+  /* Add text description */
+  for (UnsignedInteger k = 0; k < values.getDimension(); ++k)
+  {
+    data(k, 0) = (k + 1.0) + 0.08;
+    data(k, 1) = values[k];
+  }
+
+  Description names(values.getDimension());
+  for(UnsignedInteger i = 0; i < values.getDimension(); ++i)
+  {
+    OSS oss;
+    oss << "X" << i + 1;
+    names[i] = String(oss);
+  }
+
+  Text text(data, names, "right");
+  text.setColor("black");
+  graph.add(text);
+
+  // Set bounding box
```

Comment:
Could'nt this be done using `Graph::setXMargin()` and `Graph::setYMargin()`?

Reply:
Done